presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,264 @@
+        createClients();
+    }
+
+    public ElasticsearchConnectorConfig getConfig()
+    {
+        return config;
+    }
+
+    public ImmutableMap<String, Client> getClients()
+    {
+        return ImmutableMap.copyOf(clients);
+    }
+
+    public List<String> listSchemas()
+    {
+        ImmutableSet.Builder<String> builder = ImmutableSet.builder();
+        for (SchemaTableName tableName : tableDescriptions.keySet()) {
+            builder.add(tableName.getSchemaName());
+        }
+        return ImmutableList.copyOf(builder.build());
+    }
+
+    public List<SchemaTableName> listTables(String schemaNameOrNull)
+    {
+        ImmutableList.Builder<SchemaTableName> builder = ImmutableList.builder();
+        for (SchemaTableName tableName : tableDescriptions.keySet()) {
+            if (schemaNameOrNull == null || tableName.getSchemaName().equals(schemaNameOrNull)) {
+                builder.add(tableName);
+            }
+        }
+
+        return builder.build();
+    }
+
+    public ElasticsearchTableDescription getTable(String schemaName, String tableName)
+    {
+        requireNonNull(schemaName, "schemaName is null");
+        requireNonNull(tableName, "tableName is null");
+        try {
+            for (Map.Entry<SchemaTableName, ElasticsearchTableDescription> entry : tableDescriptions.entrySet()) {
+                buildColumns(entry.getKey(), entry.getValue());
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, e.getMessage());
+        }
+        return tableDescriptions.get(new SchemaTableName(schemaName, tableName));
+    }
+
+    public ClusterSearchShardsResponse getSearchShards(ElasticsearchTableDescription tableDescription)
+    {
+        Client client = clients.get(tableDescription.getClusterName());
+        ClusterSearchShardsRequest request = new ClusterSearchShardsRequest(tableDescription.getIndex());
+        try {
+            return client.admin().cluster().searchShards(request).get();
+        }
+        catch (InterruptedException | ExecutionException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, e.getMessage());
+        }
+    }
+
+    private void buildColumns(SchemaTableName schemaTableName, ElasticsearchTableDescription tableDescription)
+            throws ExecutionException, InterruptedException, IOException, JSONException
+    {
+        Set<ElasticsearchColumn> columns = new HashSet();
+
+        Client client = clients.get(tableDescription.getClusterName());
+        GetMappingsRequest mappingsRequest = new GetMappingsRequest().types(tableDescription.getType());
+        String index = tableDescription.getIndex();
+
+        if (index != null && !index.isEmpty()) {
+            mappingsRequest.indices(index);
+        }
+        ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = client.admin().indices().getMappings(mappingsRequest).get().getMappings();
```

Comment:
again there is a `Future.get()` call without a timeout here, we should definitely use a timeout.

Reply:
Also the RPC calls in this class should be wrapped in retry loops for better resilience (unless there are retries within the es library).